std/async/pool.ts

```diff
@@ -27,20 +32,39 @@ export function pooledMap<T, R>(
   (async (): Promise<void> => {
     const writer = res.writable.getWriter();
     const executing: Array<Promise<unknown>> = [];
-    for await (const item of array) {
-      const p = Promise.resolve().then(() => iteratorFn(item));
-      writer.write(p);
-      const e: Promise<unknown> = p.then(() =>
-        executing.splice(executing.indexOf(e), 1)
-      );
-      executing.push(e);
-      if (executing.length >= poolLimit) {
-        await Promise.race(executing);
+    try {
+      for await (const item of array) {
+        const p = Promise.resolve().then(() => iteratorFn(item));
+        // Only write on success. If we `writer.write()` a rejected promise,
+        // that will end the iteration. We don't want that yet. Instead let it
+        // fail the race, taking us to the catch block where all currently
+        // executing jobs are allowed to finish and all rejections among them
+        // can be reported together.
+        p.then((v) => writer.write(Promise.resolve(v))).catch(() => {});
+        const e: Promise<unknown> = p.then(() =>
+          executing.splice(executing.indexOf(e), 1)
+        );
+        executing.push(e);
+        if (executing.length >= poolLimit) {
+          await Promise.race(executing);
+        }
+      }
+      // Wait until all ongoing events have processed, then close the writer.
+      await Promise.all(executing);
+      writer.close();
+    } catch {
+      const errors = [];
+      for (const result of await Promise.allSettled(executing)) {
+        if (result.status == "rejected") {
+          errors.push(result.reason);
+        }
       }
+      const p = Promise.reject(
+        new AggregateError(errors, "Threw while mapping."),
+      );
+      p.catch(() => {});
```

Comment:
I think we shouldn't attach catch handler here because this error can be handled by the user of pooledMap.

Reply:
When we `writer.write()` this rejected promise, it causes the iterator to throw it to the user. An empty catch handler because it is treated like an unhandled rejection otherwise.

Reply:
I commented out this line and tried the script below:
```ts
import { pooledMap } from "./std/async/pool.ts";
let iter = pooledMap(4, [...Array(100)], () => {
  throw new Error();
});
try {
  for await (const _ of iter) {}
} catch (e) {
  console.log(e.errors);
}
console.log("finish");
```
and this script seems finishing without unhandled rejection error. Does this mean the error can be handled by the user?

On the other hand, I confirmed that the empty handler at the next line is necessary to avoid unhandled rejection error.

Reply:
You're right, `writer.write()`'s implementation properly inherits the rejection handling so it wouldn't be needed. I think I tested it before I added the handler below.